libs/nest/swagger/src/lib/documentation.decorator.ts

```diff
@@ -0,0 +1,93 @@
+import { applyDecorators, HttpCode } from '@nestjs/common'
+import {
+  ApiBadRequestResponse,
+  ApiConflictResponse,
+  ApiCreatedResponse,
+  ApiForbiddenResponse,
+  ApiInternalServerErrorResponse,
+  ApiNoContentResponse,
+  ApiNotFoundResponse,
+  ApiOkResponse,
+  ApiOperation,
+  ApiParam,
+  ApiQuery,
+  ApiUnauthorizedResponse,
+  ApiQueryOptions,
+  ApiResponseMetadata,
+  ApiParamOptions,
+} from '@nestjs/swagger'
+
+import { HttpProblemResponse } from '@island.is/nest/problem'
+
+interface Options {
+  isAuthorized?: boolean
+  queries?: ApiQueryOptions[]
+  params?: ApiParamOptions[]
+  description?: string
+  response: {
+    status: 200 | 201 | 204
+    type?: ApiResponseMetadata['type']
+  }
+}
+
+const getResponseDecorator = (
+  response: Options['response'],
+): MethodDecorator => {
+  switch (response.status) {
+    case 200:
+      return ApiOkResponse(response)
+    case 201:
+      return ApiCreatedResponse(response)
+    case 204:
+      return ApiNoContentResponse(response)
+  }
+}
+
+const getQueriesDecorators = (
+  queries: Options['queries'],
+): MethodDecorator[] => {
+  if (!queries) {
+    return []
+  }
+
+  return queries.map((options) => ApiQuery(options))
+}
+
+const getParamsDecorators = (params: Options['params']): MethodDecorator[] => {
+  if (!params) {
+    return []
+  }
+
+  return params.map((options) => ApiParam(options))
+}
+
+const getOptionalDecorators = (
+  isAuthorized: Options['isAuthorized'],
+): MethodDecorator[] => {
+  if (!isAuthorized) {
+    return []
+  }
+
+  return [ApiUnauthorizedResponse({ type: HttpProblemResponse })]
+}
+
+export const Documentation = ({
+  isAuthorized = true,
+  response,
+  queries,
+  params,
+  description,
+}: Options): MethodDecorator =>
+  applyDecorators(
+    HttpCode(response.status),
+    ApiNotFoundResponse,
+    ApiInternalServerErrorResponse(),
+    ApiBadRequestResponse({ type: HttpProblemResponse }),
+    ApiForbiddenResponse({ type: HttpProblemResponse }),
+    ApiConflictResponse({ type: HttpProblemResponse }),
```

Comment:
Should all operations list all of these responses even though they are never returned?

Reply:
Note that NotFoundResponse and InternalServerErrorResponse should also include a HttpProblemResponse body.

Reply:
> Should all operations list all of these responses even though they are never returned?

Nah, http conflict does not always apply. But I don't see anything wrong with always applying not found, bad request, forbidden, 500.

Reply:
See my review comment about the not found and forbidden responses.